PR: https://github.com/odoo/odoo/pull/129691

From: 35d8048044cfbef58fbf00848acf5da76f1bf7eb
From: aliya
Date: 2023-07-26 08:48:54

Structural Changes: 2
Total Changes: 37

[FIX] account_peppol: fix saving of the migration key

`account_peppol_migration_key` is a related field, stored on the `account_edi_proxy_client.user`.
We want to allow users to provide a migration key when they want to register as a new peppol participant,
migrating their existing peppol registration from another smp.
However, before they save and validate their registration,
there is no `account_edi_proxy_client.user` associated with the company and
hence the `migration_key` is never saved and never sent to the IAP server.
The solution in master will be to move this field to `res.company`
but in stable the workaround is to store it in `ir.config_parameter` and make it a computed field,
where the config parameter is updated in the inverse method on `account_peppol_migration_key`.

closes odoo/odoo#129691

X-original-commit: d5edca101e6a1dd751a3c95d713831e99fc18462
Signed-off-by: Laurent Smet (las) <las@odoo.com>
Signed-off-by: Aliya Tastemirova (alta) <alta@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account_peppol/models/account_edi_proxy_user.py
+++ b/addons/account_peppol/models/account_edi_proxy_user.py
@@ -3,7 +3,7 @@
 
 import logging
 
-from odoo import _, fields, models, tools
+from odoo import _, fields, models, modules, tools
 from odoo.exceptions import UserError
 from odoo.addons.account_edi_proxy_client.models.account_edi_proxy_user import AccountEdiProxyError
 
@@ -34,7 +34,17 @@ class AccountEdiProxyClientUser(models.Model):
                 and not self.active
                 and not self.company_id.account_edi_proxy_client_ids.filtered(lambda u: u.proxy_type == 'peppol')
             ):
-                self.company_id.account_peppol_proxy_state = 'not_registered'
+                self.company_id.write({
+                    'account_peppol_proxy_state': 'not_registered',
+                    'is_account_peppol_participant': False,
+                })
+                self.env['ir.config_parameter'].set_param(
+                    f'account_peppol.migration_key_{self.company_id.id}',
+                    False
+                )
+                # commit the above changes before raising below
+                if not tools.config['test_enable'] and not modules.module.current_test:
+                    self.env.cr.commit()
             raise AccountEdiProxyError(e.code, e.message)
         return result
 

--- a/addons/account_peppol/models/res_config_settings.py
+++ b/addons/account_peppol/models/res_config_settings.py
@@ -27,7 +27,12 @@ class ResConfigSettings(models.TransientModel):
         string="Warning",
         compute="_compute_account_peppol_endpoint_warning",
     )
-    account_peppol_migration_key = fields.Char(related='account_peppol_edi_user.peppol_migration_key', readonly=False)
+    # to be changed in master to be a related field on res_company
+    account_peppol_migration_key = fields.Char(
+        compute="_compute_account_peppol_migration_key",
+        inverse="_inverse_account_peppol_migration_key",
+        readonly=False,
+    )
     account_peppol_phone_number = fields.Char(related='company_id.account_peppol_phone_number', readonly=False)
     account_peppol_proxy_state = fields.Selection(related='company_id.account_peppol_proxy_state', readonly=False)
     account_peppol_purchase_journal_id = fields.Many2one(related='company_id.peppol_purchase_journal_id', readonly=False)
@@ -54,7 +59,7 @@ class ResConfigSettings(models.TransientModel):
 
     def _call_peppol_proxy(self, endpoint, params=None, edi_user=None):
         if not edi_user:
-            edi_user = self.company_id.account_edi_proxy_client_ids[0]
+            edi_user = self.company_id.account_edi_proxy_client_ids.filtered(lambda u: u.proxy_type == 'peppol')
 
         params = params or {}
         try:
@@ -103,6 +108,20 @@ class ResConfigSettings(models.TransientModel):
                 config.account_peppol_endpoint_warning = _("The endpoint number might not be correct. "
                                                            "Please check if you entered the right identification number.")
 
+    @api.depends('company_id')
+    def _compute_account_peppol_migration_key(self):
+        for config in self:
+            config.account_peppol_migration_key = self.env['ir.config_parameter'].get_param(
+                f'account_peppol.migration_key_{config.company_id.id}'
+            )
+
+    def _inverse_account_peppol_migration_key(self):
+        for config in self:
+            self.env['ir.config_parameter'].set_param(
+                f'account_peppol.migration_key_{config.company_id.id}',
+                config.account_peppol_migration_key
+            )
+
     # -------------------------------------------------------------------------
     # BUSINESS ACTIONS
     # -------------------------------------------------------------------------
